packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+     * @param _subject subject address for which tokens are getting deposited.
+     * @param _amount Amount of tokens getting deposited.
+     * @param _lockPeriodInSec Lock period for the tokens.
+     * @param _beneficiary Address of the beneficiary for the lock.
+     * @param _isBuy Whether the lock is created from a buy operation.
+     */
+    function _createLock(
+        address _subject,
+        uint256 _amount,
+        uint256 _lockPeriodInSec,
+        address _beneficiary,
+        bool _isBuy
+    )
+        internal
+        returns (IERC20Extended _subjectToken, uint256 unlockTimeInSec_)
+    {
+        if (_isZeroAddress(_subject)) {
+            revert Staking_InvalidSubject();
+        }
+        if (_amount == 0) {
+            revert Staking_AmountShouldBeGreaterThanZero();
+        }
+        if (_isZeroAddress(_beneficiary)) {
+            revert Staking_InvalidBeneficiary();
+        }
+        _subjectToken = IERC20Extended(tokenManager.tokens(_subject));
+        if (address(_subjectToken) == address(0)) {
+            revert Staking_InvalidSubjectToken();
+        }
+        uint256 _index = lockCount++;
+        unlockTimeInSec_ = block.timestamp + _lockPeriodInSec;
+        LockInfo memory lockInfo = LockInfo({
+            amount: _amount,
+            unlockTimeInSec: unlockTimeInSec_,
+            subject: _subject,
+            subjectToken: address(_subjectToken),
+            user: _beneficiary,
+            lockPeriodInSec: _lockPeriodInSec
+        });
+        // lock the tokens
+        locks[_index] = lockInfo;
+
+        emit Lock(
+            _beneficiary,
+            _subject,
+            address(_subjectToken),
+            _index,
+            _amount,
+            unlockTimeInSec_,
+            _lockPeriodInSec,
+            _isBuy
+        );
+    }
+
+    /**
+     * @notice Extracts expired locks and deletes them.
+     * @param _subject Subject address for which locks are being extracted.
+     * @param _indexes Indexes of the locks to be extracted.
+     * @return subjectToken_ Address of the subject token.
+     * @return totalAmount_ Total amount of tokens withdrawn.
+     */
+    function _extractExpiredAndDeleteLocks(
+        address _subject,
+        uint256[] memory _indexes
+    ) internal returns (address subjectToken_, uint256 totalAmount_) {
+        if (_isZeroAddress(_subject)) {
+            revert Staking_InvalidSubject();
+        }
+
+        if (_indexes.length == 0) {
+            revert Staking_EmptyIndexes();
+        }
+        LockInfo memory lockInfo = locks[_indexes[0]];
+        subjectToken_ = lockInfo.subjectToken;
+
+        for (uint256 i = 0; i < _indexes.length; i++) {
+            uint256 index = _indexes[i];
+            lockInfo = locks[index];
+            if (lockInfo.subject != _subject) {
+                revert Staking_SubjectsDoesNotMatch(index);
+            }
+            if (lockInfo.unlockTimeInSec > block.timestamp) {
```

Comment:
actually this is in epoch and not in sec . may be we can rename unblockTimestamp / unblockTimeEpoch

Reply:
block time is calculated in secs.